3/classification/training.py

```diff
@@ -0,0 +1,114 @@
+import numpy as np
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.nn.utils.rnn import pad_sequence
+from tqdm import tqdm_notebook as tqdm
+
+from classification.models import SpeakerIdentityLSTM, Autoencoder
+
+
+def _fix_random_state(seed=7):
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+
+
+def pack_feature_class_lists(features, classes, ids, device):
+    ids = sorted(ids, key=lambda ind: features[ind].size(0), reverse=True)
+    lengths = list(map(lambda ind: features[ind].size(0), ids))
+    seq = list(map(lambda ind: features[ind], ids))
+    inputs = pad_sequence(seq, batch_first=True).to(device)
+    targets = torch.tensor(list(map(lambda ind: classes[ind], ids)), dtype=torch.long)
+    return inputs, lengths, targets
+
+
+def train_lstm(train_features, train_classes, test_features, test_classes, num_people, device):
+    EPOCHS = 5000
+    BATCH_SIZE = 1024
+    HIDDEN_DIM = 64
+    _fix_random_state()
+
+    model = SpeakerIdentityLSTM(train_features[0].shape[1], HIDDEN_DIM, num_people, device)
+    criterion = nn.NLLLoss()
+    optimizer = optim.SGD(model.parameters(), lr=1e-2)
+
+    N_train = len(train_features)
+    N_test = len(test_features)
+    train_ids = list(range(N_train))
+    test_ids = list(range(N_test))
+
+    train_losses = []
+    test_accuracies = []
+
+    for _ in tqdm(range(EPOCHS)):
```

Comment:
Consider `trange`

Reply:
It is good to know about it. But tqdm does not work well with jupyter use tqdm_notebook.